File: sdk/javascript/examples/workflow.ts
/**
 * Example: Create a harvest event, submit for verification, verify, and publish.
 *
 * Usage:
 *   npx tsx examples/workflow.ts
 */

import { KokonutClient } from '../src/index.js';
import type { HarvestEvent } from '../src/index.js';

const client = new KokonutClient('http://localhost:8055', {
  token: process.env.KOKONUT_TOKEN,
});

const STATE_TRANSITIONS: Record<HarvestEvent['status'], HarvestEvent['status']> = {
  draft: 'submitted',
  submitted: 'verified',
  verified: 'published',
  published: 'published',
  rejected: 'submitted',
};

async function transitionHarvest(
  harvestId: string,
  targetState: HarvestEvent['status']
): Promise<HarvestEvent> {
  const current = await client.harvestEvents.get(harvestId);
  console.log(`  Current state: ${current.status}`);

  if (current.status === targetState) {
    console.log(`  Already in target state: ${targetState}`);
    return current;
  }

  const expectedNext = STATE_TRANSITIONS[current.status];
  if (expectedNext !== targetState) {
    throw new Error(`Invalid transition: ${current.status} -> ${targetState} (expected ${expectedNext})`);
  }

  const updated = await client.harvestEvents.update(harvestId, {
    status: targetState,
  });
  console.log(`  Transitioned to: ${updated.status}`);
  return updated;
}

async function main() {
  await client.login('admin@example.com', 'password123');

  const cropCycleId = 'crop-cycle-id';
  const plotId = 'plot-id';

  console.log('1. Creating draft harvest event...');
  const harvest = await client.harvestEvents.create({
    crop_cycle_id: cropCycleId,
    plot_id: plotId,
    harvest_date: new Date().toISOString(),
    quantity: 1500,
    unit: 'kg',
    quality_grade: 'A',
    gross_weight: 1500,
    net_weight: 1420,
    loss_quantity: 80,
    loss_reason: 'moisture_loss_during_drying',
    status: 'draft',
    recorded_by: 'field-supervisor-001',
  });
  console.log(`  Harvest created: ${harvest.id}\n`);

  console.log('2. Submitting harvest data...');
  await transitionHarvest(harvest.id, 'submitted');
  await client.harvestEvents.update(harvest.id, {
    quantity: 1420,
    loss_quantity: 80,
  });
  console.log('');

  console.log('3. Verifying harvest...');
  await transitionHarvest(harvest.id, 'verified');
  await client.harvestEvents.update(harvest.id, {
    status: 'verified',
  });
  console.log('');

  console.log('4. Publishing verified harvest...');
  await transitionHarvest(harvest.id, 'published');
  await client.harvestEvents.update(harvest.id, {
    status: 'published',
  });
  console.log('');

  console.log('5. Recording corresponding sale...');
  const sale = await client.salesEvents.create({
    crop_cycle_id: cropCycleId,
    harvest_event_id: harvest.id,
    buyer_name: 'Local Cooperative',
    sale_date: new Date().toISOString(),
    quantity: 1420,
    unit: 'kg',
    price_per_unit: 8.5,
    total_amount: 12070,
    currency: 'USD',
    payment_status: 'pending',
    status: 'draft',
  });
  console.log(`  Sale created: ${sale.id}`);

  console.log('\n6. Creating attestation record...');
  const attestation = await client.attestations.create({
    entity_type: 'harvest_event',
    entity_id: harvest.id,
    claim_type: 'verified_harvest',
    claim_data: {
      crop_cycle_id: cropCycleId,
      quantity: 1420,
      unit: 'kg',
      quality_grade: 'A',
      loss_rate_pct: 5.33,
      verified_by: 'field-supervisor-001',
    },
    status: 'draft',
  });
  console.log(`  Attestation created: ${attestation.id}`);

  console.log('\nWorkflow complete!');
}

main().catch(console.error);

File: sdk/python/examples/workflow.py
"""Example: Full harvest lifecycle (create -> submit -> verify -> publish).

Usage:
    KOKONUT_TOKEN=... python examples/workflow.py
"""

import os
import sys
from datetime import datetime, timezone

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from kokonut import KokonutClient

BASE_URL = os.environ.get("KOKONUT_BASE_URL", "http://localhost:8055")
TOKEN = os.environ.get("KOKONUT_TOKEN")

VALID_TRANSITIONS = {
    "draft": "submitted",
    "submitted": "verified",
    "verified": "published",
    "rejected": "submitted",
}


def transition_harvest(client: KokonutClient, harvest_id: str, target_status: str) -> dict:
    current = client.harvest_events.get(harvest_id)
    print(f"  Current state: {current.get('status')}")

    if current.get("status") == target_status:
        print(f"  Already in target state: {target_status}")
        return current

    expected = VALID_TRANSITIONS.get(current.get("status", ""))
    if expected != target_status:
        raise ValueError(
            f"Invalid transition: {current.get('status')} -> {target_status} "
            f"(expected -> {expected})"
        )

    updated = client.harvest_events.update(harvest_id, {"status": target_status})
    print(f"  Transitioned to: {updated.get('status')}")
    return updated


def main() -> None:
    client = KokonutClient(BASE_URL, token=TOKEN)

    if not TOKEN:
        client.login("admin@example.com", os.environ.get("ADMIN_PASSWORD", "changeme"))

    # These would be real UUIDs from your platform
    crop_cycle_id = "PLACEHOLDER_CROP_CYCLE_ID"
    plot_id = "PLACEHOLDER_PLOT_ID"
    location_id = "PLACEHOLDER_LOCATION_ID"

    # Step 1: Create draft harvest event
    print("1. Creating draft harvest event...")
    now = datetime.now(timezone.utc).isoformat()
    harvest = client.harvest_events.create({
        "crop_cycle_id": crop_cycle_id,
        "plot_id": plot_id,
        "location_id": location_id,
        "harvest_date": now,
        "quantity": 1500,
        "unit": "kg",
        "quality_grade": "A",
        "loss_amount": 80,
        "loss_unit": "kg",
        "loss_reason": "moisture_loss_during_drying",
        "status": "draft",
        "notes": "Field harvest from Plot A",
    })
    print(f"  Harvest created: {harvest['id']}\n")

    # Step 2: Submit after field data entry is complete
    print("2. Submitting harvest data...")
    transition_harvest(client, harvest["id"], "submitted")
    client.harvest_events.update(harvest["id"], {
        "quantity": 1420,
        "loss_amount": 80,
    })
    print("  Updated quantity to net weight\n")

    # Step 3: Submit for verification
    print("3. Submitting for verification...")
    transition_harvest(client, harvest["id"], "verified")
    client.harvest_events.update(harvest["id"], {
        "verified_by": "field-supervisor-001",
    })
    print("  Recorded verifier\n")

    # Step 4: Publish
    print("4. Publishing verified harvest...")
    transition_harvest(client, harvest["id"], "published")
    print("  Harvest is now publicly visible\n")

    # Step 5: Record corresponding sale
    print("5. Recording corresponding sale...")
    sale = client.sales_events.create({
        "crop_cycle_id": crop_cycle_id,
        "harvest_id": harvest["id"],
        "location_id": location_id,
        "buyer": "Local Cooperative",
        "buyer_type": "cooperative",
        "sale_date": now,
        "quantity": 1420,
        "unit": "kg",
        "price_per_unit": 8.50,
        "total_amount": 12070,
        "currency": "USD",
        "payment_status": "pending",
        "status": "draft",
    })
    print(f"  Sale created: {sale['id']}\n")

    # Step 6: Create attestation record
    print("6. Creating attestation record...")
    attestation = client.attestations.create({
        "schema_id": "PLACEHOLDER_SCHEMA_ID",
        "subject_type": "harvest_event",
        "subject_id": harvest["id"],
        "claim_type": "verified_harvest",
        "claim_data": {
            "crop_cycle_id": crop_cycle_id,
            "quantity": 1420,
            "unit": "kg",
            "quality_grade": "A",
            "loss_rate_pct": 5.33,
            "verified_by": "field-supervisor-001",
        },
        "status": "draft",
    })
    print(f"  Attestation created: {attestation['id']}")

    print("\nWorkflow complete!")
    print(f"  Harvest:  {harvest['id']}")
    print(f"  Sale:     {sale['id']}")
    print(f"  Attestation: {attestation['id']}")


if __name__ == "__main__":
    main()
